hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/AuthenticationBaseMessageTask.java

```diff
@@ -76,65 +76,80 @@ private void handleEndpointNotCreatedConnectionNotAlive() {
 
     @Override
     public Object call() {
-        boolean authenticated = authenticate();
-
-        if (authenticated) {
-            return handleAuthenticated();
+        byte serializationServiceVersion = serializationService.getVersion();
+        AuthenticationStatus authenticationStatus;
+        if (clientSerializationVersion != serializationServiceVersion) {
+            authenticationStatus = AuthenticationStatus.SERIALIZATION_VERSION_MISMATCH;
         } else {
-            return handleUnauthenticated();
+            authenticationStatus = authenticate();
+        }
+        switch (authenticationStatus) {
+            case AUTHENTICATED:
+                return handleAuthenticated();
+            case CREDENTIALS_FAILED:
+                return handleUnauthenticated();
+            case SERIALIZATION_VERSION_MISMATCH:
+                return handleSerializationVersionMismatch();
+            default:
         }
+        throw new IllegalStateException("Execution cannot reach here.");
```

Comment:
Place this exception in the default: clause. There is no really a point in providing an empty default class since that is the 'default' behavior anyway.